### PR TITLE
Make web server settings configurable

### PR DESCRIPTION
**Pull request by MebinAbraham**
### What is the context of this PR?
Makes the web server settings configurable so that we can't easily swap out web server/settings as well as enable and disable the use of GRPC for Datastore.
Because it seemed appropriate, I have renamed `GUNICORN_KEEP_ALIVE` to `HTTP_KEEP_ALIVE` to be more explicit.

### How to review 
Ensure:
- app works using the different web server commands in Makefile
- the env vars are actually being applied as expected.
- helm deployment of the app works as expected. Ensure you can changes settings as expected. You can use https://github.com/ONSdigital/eq-pipelines/pull/87 to test.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

### CODE REVIEW

File: .development.env [hunk 1]
```diff
@@ -16,7 +16,7 @@ EQ_NEW_RELIC_ENABLED=False
 EQ_SECRETS_FILE=dev-secrets.yml
 EQ_KEYS_FILE=dev-keys.yml
 AWS_DEFAULT_REGION=eu-west-1
-GUNICORN_KEEP_ALIVE=2
+HTTP_KEEP_ALIVE=2
```
Comment (JamesGardiner) [inline .development.env:19]: Is this a TCP or HTTP keep alive?

File: Dockerfile [hunk 2]
```diff
@@ -7,12 +7,13 @@ RUN apt update && apt install -y curl unzip libsnappy-dev build-essential jq
...
 
-ENV WEB_SERVER gunicorn
+ENV WEB_SERVER_TYPE gunicorn
+ENV WEB_SERVER_WORKER_CLASS gevent
 ENV WEB_SERVER_WORKERS 3
-ENV GUNICORN_KEEP_ALIVE 2
-ENV GUNICORN_CMD_ARGS -c gunicorn_config.py
 ENV WEB_SERVER_THREADS 10
 ENV WEB_SERVER_ASYNC_CORES 10
+ENV HTTP_KEEP_ALIVE 2
```
Comment (JamesGardiner) [inline Dockerfile:15]:
```suggestion
ENV WEB_SERVER_KEEP_ALIVE 2
```

I don't think we should make an assumption about what protocol is being used for the keepalive

File: .development.env (hunk 1 above)
Comment (JamesGardiner) [inline .development.env:19]:
```suggestion
WEB_SERVER_KEEP_ALIVE=2
```

I don't think we should make an assumption about what protocol is being used for the keepalive

**Review by JamesGardiner — CHANGES_REQUESTED** [review]
Also needs the gRPC flag to be configurable 👍

File: Makefile [hunk 3]
```diff
@@ -55,17 +55,20 @@ link-development-env:
 run: build link-development-env
 	pipenv run flask run
 
-run-gunicorn: link-development-env
-	WEB_SERVER=gunicorn pipenv run ./run_app.sh
+run-gunicorn-gevent: link-development-env
+	WEB_SERVER_TYPE=gunicorn WEB_SERVER_WORKER_CLASS=gevent pipenv run ./run_app.sh
+
+run-gunicorn-gthread: link-development-env
```
Comment (JamesGardiner) [inline Makefile:61]: What do you think about making an assumption that gRPC will always be on when using gthread?

**Comment by MebinAbraham** [comment]
> Also needs the gRPC flag to be configurable 👍

Done

File: Dockerfile (hunk 2 above)
Comment (MebinAbraham) [reply Dockerfile:15]: I did HTTP to make it explicit. For uwsgi, it has http_keepalive and so_keepalive for sockets.

File: .development.env (hunk 1 above)
Comment (MebinAbraham) [reply .development.env:19]: https://github.com/ONSdigital/eq-questionnaire-runner/pull/314#discussion_r496597536

File: Makefile (hunk 3 above)
Comment (MebinAbraham) [reply Makefile:61]: Could do, what about uwsgi?. Also, for dev, it shouldn't make much of a difference.

Comment (JamesGardiner) [reply Makefile:61]: uWSGI as well

File: app/setup.py [hunk 4]
```diff
@@ -279,7 +279,7 @@ def setup_dynamodb(application):
 
 
 def setup_datastore(application):
-    client = datastore.Client(_use_grpc=False)
+    client = datastore.Client(_use_grpc=application.config["DATASTORE_USE_GRPC"])
```
Comment (ajmaddaford) [inline app/setup.py:282]: There is an environment variable (GOOGLE_CLOUD_DISABLE_GRPC) used by the library that we could consider using - https://github.com/googleapis/python-datastore/blob/master/google/cloud/datastore/client.py#L60

Although, that is misleading if the PubSub library ignores it 🤔

File: Makefile [hunk 5]
```diff
@@ -55,17 +55,20 @@ link-development-env:
...
-	WEB_SERVER=uwsgi pipenv run ./run_app.sh
+	WEB_SERVER_TYPE=uwsgi pipenv run ./run_app.sh
 
 run-uwsgi-threads: link-development-env
-	WEB_SERVER=uwsgi-threads pipenv run ./run_app.sh
+	WEB_SERVER_TYPE=uwsgi-threads pipenv run ./run_app.sh
 
 run-uwsgi-async: link-development-env
-	WEB_SERVER=uwsgi-async pipenv run ./run_app.sh
+	WEB_SERVER_TYPE=uwsgi-async pipenv run ./run_app.sh
```
Comment (ajmaddaford) [inline Makefile:71]: The run commands for gunicorn and uwsgi are inconsistent. We should consider `WEB_SERVER_TYPE` of `gunicorn-threads` and `gunicorn-async`.

File: k8s/helm/values.yaml [hunk 6]
```diff
@@ -13,6 +13,16 @@ googleTagManagerId: ""
 googleTagManagerAuth: ""
 googleTagManagerPreview: ""
 
+webServer:
+  type: "gunicorn"
+  workerClass: "gevent"
+  workers: "7"
+  threads: "10"
+  asyncCores: "10"
```
Comment (ajmaddaford) [inline k8s/helm/values.yaml:21]: If this only applies to uwsgi, this should be named something like `uwsgiAsyncCores`.

File: app/setup.py (hunk 4 above)
Comment (MebinAbraham) [reply app/setup.py:282]: Myself and James did discuss this env var previous;y but chose to use our own one to have finer control on which lib has grpc enabled etc.

File: Makefile (hunk 5 above)
Comment (MebinAbraham) [reply Makefile:71]: Fair point however, we would still need `worker_class = os.getenv("WEB_SERVER_WORKER_CLASS")` given we use a py config file and not cmd line args.

Comment (ajmaddaford) [reply Makefile:71]: I was thinking the env var would be set in the run_app script.

File: k8s/helm/values.yaml (hunk 6 above)
Comment (ajmaddaford) [reply k8s/helm/values.yaml]: Sorry, I should have been clearer. This should be changed everywhere e.g. the env var should be `WEB_SERVER_UWSGI_ASYNC_CORES` (or `UWSGI_ASYNC_CORES`).

File: Makefile (hunk 5 above)
Comment (MebinAbraham) [reply Makefile]: Update webserver type to be consistent and to use env var in run_app script.

File: k8s/helm/values.yaml (hunk 6 above)
Comment (MebinAbraham) [reply k8s/helm/values.yaml]: I did wonder 😄 
I went with `UWSGI_ASYNC_CORES` because we have `GUNICORN_CMD_ARGS`.

File: run_app.sh [hunk 7]
```diff
@@ -2,16 +2,16 @@
...
+    run_command="WEB_SERVER_WORKER_CLASS=gthread gunicorn application:application"
+elif [ "$WEB_SERVER_TYPE" = "uwsgi" ]; then
     run_command="uwsgi uwsgi.ini --workers ${WEB_SERVER_WORKERS}"
-elif [ "$web_server" = "uwsgi-threads" ]; then
+elif [ "$WEB_SERVER_TYPE" = "uwsgi-threads" ]; then
     run_command="uwsgi uwsgi.ini --workers ${WEB_SERVER_WORKERS} --enable-threads --threads ${WEB_SERVER_THREADS}"
-elif [ "$web_server" = "uwsgi-async" ]; then
-    run_command="uwsgi uwsgi.ini --module patched:application --workers ${WEB_SERVER_WORKERS} --single-interpreter --gevent ${WEB_SERVER_ASYNC_CORES}"
+elif [ "$WEB_SERVER_TYPE" = "uwsgi-async" ]; then
+    run_command="uwsgi uwsgi.ini --module patched:application --workers ${WEB_SERVER_WORKERS} --single-interpreter --gevent ${UWSGI_ASYNC_CORES}"
```
Comment (ajmaddaford) [inline run_app.sh:14]: I think we need a different environment variable name. uWSGI is parsing this and reading it as a value for an unknown directive `async-cores`:
```
[strict-mode] unknown config directive: async-cores
```

Comment (MebinAbraham) [reply run_app.sh]: Have renamed `UWSGI_ASYNC_CORES` to `WEB_SERVER_UWSGI_ASYNC_CORES`

**Review by ajmaddaford — APPROVED** [review]

**Review by JamesGardiner — APPROVED** [review]

**Comment by pricem14pc** [comment]
Do we need to add these to the list of environment variable params in deploy_app.yaml to allow them to be overridden?

```
  CDN_URL:
  WEB_SERVER_TYPE:
  WEB_SERVER_WORKERS:
  WEB_SERVER_THREADS:
  WEB_SERVER_UWSGI_ASYNC_CORES:
  DATASTORE_USE_GRPC:
```

**Comment by MebinAbraham** [comment]
> Do we need to add these to the list of environment variable params in deploy_app.yaml to allow them to be overridden?
> 
> ```
>   CDN_URL:
>   WEB_SERVER_TYPE:
>   WEB_SERVER_WORKERS:
>   WEB_SERVER_THREADS:
>   WEB_SERVER_UWSGI_ASYNC_CORES:
>   DATASTORE_USE_GRPC:
> ```

The ones introduced in this PR has already been added I believe.

**Comment by MebinAbraham** [comment]
@ajmaddaford @JamesGardiner Made minor changes to the README, can you make sure you are happy with the changes. Thanks

**Review by JamesGardiner — APPROVED** [review]

**Review by ajmaddaford — APPROVED** [review]

**Pull request merged**